File: 2015/day22/fu.rb
MISSILE  = 0
DRAIN    = 1
SHIELD   = 2
POISON   = 3
RECHARGE = 4

SPELL_COSTS  = [53, 73, 113, 173, 229]
SPELL_TIMERS = [1, 1, 6, 6, 5]

def deep_dup(object)
  Marshal.load(Marshal.dump(object))
end

nodes = [
  {
    hero_hp: 50,
    hero_mana: 500,
    hero_armor: 0,
    boss_hp: 58,
    boss_damage: 9,
    mana_spent: 0,
    spell_timers: [0, 0, 0, 0, 0],
    is_hero_turn: true,
  }
]

until nodes.empty? do
  n = nodes.shift

  # reset armor
  n[:hero_armor] = 0

  # PART 2: hard mode
  if n[:is_hero_turn]
    n[:hero_hp] -= 1
    next if n[:hero_hp] <= 0
  end

  # apply spell effects
  0.upto(4) do |s|
    if n[:spell_timers][s] > 0
      n[:spell_timers][s] -= 1
      case s
      when MISSILE  then n[:boss_hp] -= 4
      when DRAIN    then n[:boss_hp] -= 2; n[:hero_hp] += 2
      when SHIELD   then n[:hero_armor] = 7
      when POISON   then n[:boss_hp] -= 3
      when RECHARGE then n[:hero_mana] += 101
      end
    end
  end

  next  if n[:hero_hp] <= 0
  break if n[:boss_hp] <= 0 # \o/

  if n[:is_hero_turn]
    0.upto(4) do |s|
      if n[:spell_timers][s] == 0 && n[:hero_mana] >= SPELL_COSTS[s]
        node = deep_dup(n)
        node[:hero_mana] -= SPELL_COSTS[s]
        node[:mana_spent] += SPELL_COSTS[s]
        node[:spell_timers][s] = SPELL_TIMERS[s]
        node[:is_hero_turn] = !node[:is_hero_turn]
        nodes << node
      end
    end
  else
    node = deep_dup(n)
    node[:hero_hp] -= [1, n[:boss_damage] - n[:hero_armor]].max
    node[:is_hero_turn] = !node[:is_hero_turn]
    nodes << node
  end
end

p n[:mana_spent]

File: 2015/day22/fu.clj
(let [boss {:hp 58 :damage 8}
      player {:hp 50 :mana 500 :mana-spent 0 :armor 0 :shield 0 :poison 0 :recharge 0}
      spells ["magic missile" "drain" "shield" "poison" "recharge"]]

  (defn apply-effects [player boss]
    [(cond-> player
       (= 0 (player :shield)) (assoc-in [:armor] 0)
       (< 0 (player :shield)) (assoc-in [:armor] 7)
       (< 0 (player :recharge)) (update-in [:mana] + 101)
       (< 0 (player :shield)) (update-in [:shield] dec)
       (< 0 (player :poison)) (update-in [:poison] dec)
       (< 0 (player :recharge)) (update-in [:recharge] dec))
     (cond-> boss (< 0 (player :poison)) (update-in [:hp] - 3))])

  (defn attack [player boss]
    (let [[p b] (apply-effects player boss)]
      [(update-in p [:hp] - (max 1 (- (b :damage) (p :armor))))
       b]))

  (defn cast-spell [player boss spell hard]
    (let [player-mod (if hard (update-in player [:hp] dec) player)
          [p b] (apply-effects player-mod boss)]
      (if (and hard (>= 0 (player-mod :hp)))
        [player-mod boss]
        (cond
          (= spell "magic missile")
            [(-> p (update-in [:mana] - 53) (update-in [:mana-spent] + 53))
             (-> b (update-in [:hp] - 4))]
          (= spell "drain")
            [(-> p (update-in [:mana] - 73) (update-in [:mana-spent] + 73) (update-in [:hp] + 2))
             (-> b (update-in [:hp] - 2))]
          (= spell "shield")
            [(-> p (update-in [:mana] - 113) (update-in [:mana-spent] + 113) (assoc-in [:shield] 6))
             b]
          (= spell "poison")
            [(-> p (update-in [:mana] - 173) (update-in [:mana-spent] + 173) (assoc-in [:poison] 6))
             b]
          (= spell "recharge")
            [(-> p (update-in [:mana] - 229) (update-in [:mana-spent] + 229) (assoc-in [:recharge] 5))
             b]))))

  (defn choose-spell [player]
    (let [choices '(0)
          mana (player :mana)]
      (nth spells (rand-nth (cond-> choices
                              (> mana 73) (conj 1)
                              (and (= 0 (player :shield)) (> mana 113)) (conj 2)
                              (and (= 0 (player :poison)) (> mana 173)) (conj 3)
                              (and (= 0 (player :recharge)) (> mana 229)) (conj 4))))))

  (defn attack-til-dead [player boss hard]
    (loop [[p b] (cast-spell player boss (choose-spell player) hard)
           turn "player"]
      (if (and (>= 0 (b :hp)) (< 0 (p :mana)) (< 0 (p :hp)))
        (p :mana-spent)
        (if (or (>= 0 (p :hp)) (>= 0 (p :mana)))
          nil
          (if (= turn "player")
            (recur (attack p b) "boss")
            (recur (cast-spell p b (choose-spell player) hard) "player"))))))

  (println "Part 1:" (apply min (filter some? (take 5000000 (repeatedly #(attack-til-dead player boss false))))))
  (println "Part 2:" (apply min (filter some? (take 5000000 (repeatedly #(attack-til-dead player boss true)))))))
